src/combinatorics/simple_graph/handshake.lean

```diff
@@ -0,0 +1,251 @@
+/-
+Copyright (c) 2020 Kyle Miller. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Kyle Miller.
+-/
+import combinatorics.simple_graph.basic
+import algebra.big_operators.basic
+import data.nat.parity
+import data.zmod.basic
+import tactic.omega
+/-!
+# Degree-sum formula and handshaking lemma
+
+The degree-sum formula is that the sum of the degrees of a finite
+graph is equal to twice the number of edges.  The handshaking lemma is
+a corollary, which is that the number of odd-degree vertices is even.
+
+## Main definitions
+
+- A `dart` is a directed edge, consisting of a vertex and an edge incident to it.
+- `simple_graph.sum_degrees_eq_twice_card_edges` is the degree-sum formula.
+- `simple_graph.card_odd_degree_vertices_is_even` is the handshaking lemma.
+- `simple_graph.card_odd_degree_vertices_ne_is_odd` is that the number of odd-degree
+  vertices different from a given odd-degree vertex is odd.
+- `simple_graph.exists_ne_odd_degree_if_exists_odd` is that the existence of an
+  odd-degree vertex implies the existence of another one.
+
+## Implementation notes
+
+We give a combinatorial proof by using the fact that the map from
+darts to vertices has fibers whose cardinalities are the degrees and
+that the map from darts to edges is 2-to-1.
+
+## Tags
+
+simple graphs, sums, degree-sum formula, handshaking lemma
+-/
+open finset
+
+open_locale big_operators
+
+namespace simple_graph
+universes u v
+variables {V : Type u} (G : simple_graph V)
+
+/-- A `dart` is a directed edge, consisting of a vertex and an edge incident to it -/
+@[ext, derive decidable_eq]
+structure dart :=
+(vert : V)
+(edge : sym2 V)
+(edge_mem : edge ∈ G.edge_set)
+(vert_mem : vert ∈ edge)
+
+/-- There is an equivalence between darts and pairs of a vertex and an incident edge. -/
+def dart_equiv_sigma : G.dart ≃ Σ v, G.incidence_set v :=
+{ to_fun := λ d, ⟨d.vert, d.edge, d.edge_mem, d.vert_mem⟩,
+  inv_fun := λ s, ⟨s.fst, s.snd, s.snd.property.1, s.snd.property.2⟩,
+  left_inv := λ d, by { ext, simp only, simp only [subtype.coe_mk], },
+  right_inv := λ s, by { ext, simp only, simp only [subtype.coe_eta], } }
+
+instance dart.fintype [fintype V] [decidable_eq V] [decidable_rel G.adj] : fintype (dart G) :=
+fintype.of_equiv _ G.dart_equiv_sigma.symm
+
+instance dart.inhabited [inhabited V] [inhabited (G.incidence_set (default _))] :
+  inhabited G.dart := ⟨G.dart_equiv_sigma.symm ⟨default _, default _⟩⟩
+
+variables {G}
+
+/-- Reverses the orientation of a dart. -/
+def dart.rev [decidable_eq V] (d : G.dart) : G.dart :=
+{ vert := sym2.mem.other' d.vert_mem,
+  edge := d.edge,
+  edge_mem := d.edge_mem,
+  vert_mem := sym2.mem_other_mem' d.vert_mem }
+
+@[simp] lemma dart_edge_of_dart_rev [decidable_eq V] (d : G.dart) : d.rev.edge = d.edge := rfl
+
+lemma dart_rev_invol [decidable_eq V] (d : G.dart) : d.rev.rev = d :=
+begin
+  apply dart.ext,
+  apply sym2.other_invol',
+  simp only [dart_edge_of_dart_rev],
+end
```

Comment:
It would be good to provide the `involutive` object too:
```suggestion
@[simp] lemma dart_rev_invol [decidable_eq V] (d : G.dart) : d.rev.rev = d :=
begin
  apply dart.ext,
  apply sym2.other_invol',
  simp only [dart_edge_of_dart_rev],
end

@[simp] lemma dart_rev_involutive [decidable_eq V] : function.involutive (dart.rev G.dart → G.dart) :=
dart_rev_invol 
```